content/post/simulating-event-windows.md

```diff
@@ -0,0 +1,171 @@
+date = 2017-11-09T00:00:00Z
+tags = [
+    "wallaroo",
+    "windowing",
+    "partitioning"
+]
+categories = [
+    "Windowing"
+]
+description = "Creating event boundaries and processing cumulative state"
+author = "amosca"
++++
+
+Certain applications lend themselves to pure parallel computation better than others. In some cases we require to apply certain algorithms over a "window" in our data. This means that after we have complete a certain amount of processing (be it time, number of messages or some other arbitraty metric), we want to perform a special action for the data in that window. An example application of this could be producing stats for log files over a certain period of time. We may want to produce our stats in the form of periodic summarier (e.g. daily), in which case the computation in Wallaroo would have to know when it has received the last message for a particular day. We are working hard every day to support new use patterns, and this type of windowing can already be simulated without any native support in Wallaroo.
+
+This blog entry is about the aforementioned example, and how this can be implemented in the current version of Wallaroo (0.2.1).
+
+## Types of windowing
+
+There are a few types of windowing that can be implemented, and it is important that we clarify the distinctions. In our example we will be focusing on event-based windowing.
+
+#### Event-based windowing
+
+This is the most arbitrary kind of windowing. The boundaries of each window are determined outside of Wallaroo, and can be triggered by any event. This is also the simplest kind of windowing to think about: when we are told that a window ends, we run work on the aggregate state and start over.
+
+#### Internally triggered windowing
+
+This is very similar to event-based windowing, except that the window is triggered from within the computation. This is usually based on the internal state of the computation, and it is the responsibility of the computation itself to determine when a window has finished and when to start a new one.
+
+#### Time-based windowing (Wall clock)
+
+A new window is started at regular intervals using a timer. This is currently not supported in Wallaroo.
+
+#### Time-based windowing (Event clock)
+
+The time element is taken from information from the message rather than a timer, and windows are created accordingly. In order to implement this type of windowing, multiple windows have to be accumulated simultaneously. It is currently not possible to implement this kind of windowing easily.
+
+## Log-file analytics application
+
+Let's return to our log file analyzer. We will assume the goal of counting different return codes on a daily basis. This is a very basic example, but it already includes the important elements needed to create a windowed application. We will stream our logfiles line-by-line to the topology, and will output one message per (day, return code) pairing. So, for instance, if we input five lines as follows
```

Comment:
I think the use "topology" here is a little confusing. Would it work with "application"?